### PR TITLE
KAFKA-10000: Exactly-once source tasks (KIP-618)

### PR DESCRIPTION
**Pull request by C0urante**
Implements a source task wrapper (`ExactlyOnceWorkerSourceTask`) that follows the behavior described in [KIP-618](https://cwiki.apache.org/confluence/display/KAFKA/KIP-618%3A+Exactly-Once+Support+for+Source+Connectors#KIP618:ExactlyOnceSupportforSourceConnectors-Atomicoffsetwrites) for writing source records and their offsets in transactions with user-configurable (and sometimes connector-defined) boundaries.

Relies on changes from:
- https://github.com/apache/kafka/pull/11772
- https://github.com/apache/kafka/pull/11775

### CODE REVIEW

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java:27]: This is a placeholder implementation. This class is fully implemented in the downstream PR https://github.com/apache/kafka/pull/11781

**Comment by C0urante** [comment]
Jenkins build failures appear to be due to the issue addressed by https://github.com/apache/kafka/pull/10702. I've pushed a change that applies the same fix from that PR to the newly-added failing test classes for this PR.

**Comment by C0urante** [comment]
Converting to draft until upstream PRs are reviewed.

**Comment by C0urante** [comment]
Given that all merge conflicts have been resolved and https://github.com/apache/kafka/pull/11775 has already been approved, marking this as ready for review.

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/OffsetStorageWriterTest.java
Comment (mimaison) [inline connect/runtime/src/test/java/org/apache/kafka/connect/storage/OffsetStorageWriterTest.java]: Can we undo this?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:1345]: Are you planning to address this `TODO` in one of the other PRs? Same below in `ExactlyOnceSourceTaskBuilder`

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Nit: can we move this block above the admin block above? so it follows the same flow than `SourceTaskBuilder.doBuild()`

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java:284]: Nit: All other messages are capitalized

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java]: Nit: Let's keep the new line

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (mimaison) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Should we use `DistributedConfig.transactionalProducerId()` here?

**Review by mimaison — COMMENTED** [review]
Thanks @C0urante for the PR. I've made a first pass over the main (I've not looked at tests yet) and it looks pretty good. I left a few minor comments.

**Comment by mimaison** [comment]
@rhauch @kkonstantine This is a pretty important feature, can you take a look? The first few PRs were relatively simple but in the remaining PRs things are getting a bit more involved so it would be good to get another pair of eyes on it.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:1345]: Yes; per-connector offsets topics are implemented in https://github.com/apache/kafka/pull/11781.

File: connect/runtime/src/test/java/org/apache/kafka/connect/storage/OffsetStorageWriterTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/storage/OffsetStorageWriterTest.java]: Apologies, will revert and remove similar changes from other PRs.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java:284]: Good point; given that `Utils::closeQuietly` uses the second parameter in the middle of a [log message](https://github.com/apache/kafka/blob/07459d2eb3e6c2278f22021854ae79c11cf6129b/clients/src/main/java/org/apache/kafka/common/utils/Utils.java#L1007) I'm inclined to lowercase each of these.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: No, the two serve different purposes.

- `DistributedConfig::transactionalProducerId` computes the transactional ID that the leader should use when writing to the config topic
- `Worker::transactionalId` computes the transactional ID that exactly-once source tasks should use when writing to Kafka

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Good idea, done.

**Comment by C0urante** [comment]
Thanks @mimaison. I've addressed your comments and this is ready for another round

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (mimaison) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Thanks, that makes sense. I thought we wanted to use `DistributedConfig::transactionalProducerId` as a prefix in `Worker::transactionalId` but that's not necessary.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (mimaison) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:1345]: Can we add a comment ? so it's clear to anyone that stumbles on this

**Comment by mimaison** [comment]
Thanks for the updates @C0urante ! I've on PTO next week, I'll take a look at the tests when I'm back.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:1345]: Ack, done.

**Comment by C0urante** [comment]
@mimaison I'm hoping you can take a look sometime this week. There's still a lot left to review for this KIP and it's worrying that the 3.3.0 feature freeze deadline is creeping closer with little progress being made.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java
Comment (tombentley) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java]: Why should they be completed by this point? I.e. what in the Java memory model guarantees it?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (tombentley) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Is this worth a comment?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (tombentley) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Perhaps we should call this `taskTransactionalId` to distinguish it from the transactional id used for the config topic?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTransactionContext.java
Comment (tombentley) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTransactionContext.java:37]: A comment about thread safety here seems warranted. It's not clear which threads are accessing an instance of quite what the mechanisms for ensuring thread safety are. E.g. why are `abortTransaction`/`commitTransaction` synchronized, but only volatile writes (surely one or other thread safety mechanism is needed, but not both)? Why is `abortTransaction` not `synchronized`?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java
Comment (tombentley) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java:902]: No verify?

**Review by tombentley — COMMENTED** [review]
Thanks @C0urante, I've made an initial pass.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: 👍 added a comment and updated the method name.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java]: That's a fair point. Right now this is a result of the threading model of the producer, which blocks on the completion of all in-flight batches and their callbacks before sending a request to the broker to end a transaction. This is not technically guaranteed by the documented behavior for `Producer::commitTransaction`, though; the only guarantees we get from the [Javadocs](https://kafka.apache.org/32/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html#commitTransaction()) are:
> This method will flush any unsent records before actually committing the transaction. Further, if any of the [send(ProducerRecord)](https://kafka.apache.org/32/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html#send(org.apache.kafka.clients.producer.ProducerRecord)) calls which were part of the transaction hit irrecoverable errors, this method will throw the last received exception immediately and the transaction will not be committed.

If we don't want to rely on this technically-undocumented behavior, we can add some more bookkeeping to ensure this ourselves. Alternatively, we can rely on this behavior and I can remove the comment above about producer callback guarantees and the bookkeeping we already do on that front (which I've done for now).

@hachikuji @guozhangwang do you have thoughts about whether it should be part of the contract for `Producer::commitTransaction` that all user-supplied callbacks for records within the transaction are fired before the method returns? The use case is tracking record metadata that's only available after the record is ack'd and then reporting it to source tasks, before clearing a local cache of that metadata.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Ack, done.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTransactionContext.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTransactionContext.java:37]: These are all good points. The lack of synchronization on `abortTransaction` is a mistake, as is declaring `batchCommitRequested` and `batchAbortRequested` volatile (redundant since they're always accessed inside synchronized blocks). Additionally, although it might help with readability, it's probably overkill to make the `checkBatchRequestsConsistency` and `checkRecordRequestConsistency` methods synchronized, since they're always called from within synchronized methods anyways.

After that, the thread safety mechanisms here become pretty simple: every public method is synchronized. I've also made a note of the thread safety expectations in the class-level Javadocs; LMKWYT.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java:902]: Good catch, added and fixed the test accordingly.

**Comment by C0urante** [comment]
Thanks Tom. Useful comments RE thread safety and an interesting point about transactional producer callbacks. Hoping we can do things the easy way on that front and let the clients library do some of the lifting for us, but if we can't settle on that by mid next week or actively decide against it before then, can do that manually in the `ExactlyOnceWorkerSourceTask` class instead.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java
Comment (tombentley) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java]: @C0urante thanks for the explanation. I don't think a revised `Producer` contract would be sufficient though. It's not the _blocking_ per-se, it's the visibility of mutations to `commitableRecords` done on those threads being correctly synchronized with reads on this line. `Producer` can't help with that because it doesn't know about `commitableRecords`, and so the writes and the read could be re-ordered anyway.

Comment (tombentley) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java]: I think all we need is a `syncrhonized(commitableRecords)` around these accesses.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTransactionContext.java
Comment (tombentley) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTransactionContext.java:37]: LGTM.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java]: Ah, fair. I was taking a charitable interpretation of the [Javadocs](https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html), which state:

> If multiple threads access a linked hash map concurrently, and at least one of the threads modifies the map structurally, it must be synchronized externally.

In this case, by the time we get to this line, we don't technically have multiple threads accessing the map concurrently; it's guaranteed that this will be the only thread reading from or writing to the map. But it costs very little to add the synchronization barrier, so I've done that.

Also emailed Jason and Guozhang for their thoughts on callbacks with transactional producers. Hoping to hear back soon 🤞

Comment (hachikuji) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java]: Yeah, I think it's a documentation gap. It is intended for callbacks to be invoked before `EndTxn` is sent. It would be a little strange otherwise. Worth noting that the callback is just for notification. Even if the callback raises an exception, we won't stop an active commit.

Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java]: Thanks Jason, filed https://issues.apache.org/jira/browse/KAFKA-13967 to track the producer docs improvement and will have a PR out for it shortly.

Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java]: Javadocs PR: https://github.com/apache/kafka/pull/12264

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java
Comment (tombentley) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java]: The Producer has it's own `TransactionManager`. Maybe we should call this `TransactionBoundaryManager` to avoid  confusion. It's also a better description of what this is seeking to abstract over imho.

**Review by tombentley — APPROVED** [review]
Thanks again @C0urante.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTask.java]: Good point, no need to complicate IDE class lookups. Changed to `TransactionBoundaryManager`

**Review by mimaison — APPROVED** [review]
LGTM, thanks @C0urante

**Comment by mimaison** [comment]
@C0urante I merged #11779 so this needs a rebase before merging

**Comment by C0urante** [comment]
Thanks Mickael, finished the rebase and handled the TODO from https://github.com/apache/kafka/pull/11779#discussion_r891379196.

**Pull request merged**
